sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala

```diff
@@ -617,10 +617,55 @@ case class DescribeTableCommand(
   }
 
   private def append(
-      buffer: ArrayBuffer[Row], column: String, dataType: String, comment: String): Unit = {
+    buffer: ArrayBuffer[Row], column: String, dataType: String, comment: String): Unit = {
     buffer += Row(column, dataType, comment)
   }
 }
+/**
+ * Command can of the following form.
+ * {{{
+ *   DESCRIBE [EXTENDED|FORMATTED] table_name partitionSpec?;
+ * }}}
+ */
+case class DescribeTableCommand(
+    tableIdent: TableIdentifier,
+    partitionSpec: TablePartitionSpec,
+    isExtended: Boolean)
+  extends DescribeTableCommandBase {
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val result = new ArrayBuffer[Row]
+    describeTable(sparkSession, tableIdent, partitionSpec, isExtended, result)
+    result
+  }
+}
+
+/**
+ * Command can of the following form.
+ * {{{
+ *   DESCRIBE [QUERY] query_statement
+ * }}}
+ */
+case class DescribeQueryCommand(query: LogicalPlan)
+  extends DescribeTableCommandBase {
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val result = new ArrayBuffer[Row]
+    query match {
+      case p if p.collectFirst { case _: InsertIntoTable | _: InsertIntoDir => true }.isDefined =>
+        throw new AnalysisException("Describe command is not supported for insert statements.")
```

Comment:
is this behavior from DB2? `sql("INSERT ...").printSchema` works and prints empty schema, shall we make the behavior consistent?

Reply:
@cloud-fan In DB2, we get an error when we describe on "insert"
```
db2 => describe insert into t1 values (1,1)
SQL0104N  An unexpected token "insert" was found following "DESCRIBE".
```
Personally, i feel its better to let users know that this command is not meant for "insert" than returning an empty schema. But if you feel otherwise, i will change. Let me know please.

Reply:
I don't have a strong preference here, but the DB2 does this in the parser, I'm wondering if we should do this in the parser as well.

instead of `(DESC | DESCRIBE) QUERY? query`, how about `(DESC | DESCRIBE) QUERY? querySpecification`?

Reply:
@cloud-fan the only thing is we may leave out a few statements if we restrict it to querySpecification like VALUES (...), WITH..  etc .. Is it okay if we create a parserRule like following ?

queryDescribe
       : stmtType1
       | stmtType2
       ;

Reply:
SGTM

Reply:
@cloud-fan I tried to add a rule, but it actually gets complicated as the rules are intertwined. In order to do this, i have to create multiple parallel rules for describequery and then implement the visit method. So lets throw analysis exception for now ?

Reply:
then how about go with `querySpecification`, to support only the `SELECT ...` query at the beginning? we can support more in the future.

Reply:
@cloud-fan OK.. let me give that a try.